syncer/sync_pipe.go

```diff
@@ -0,0 +1,671 @@
+	case del:
+		m.WithLabelValues("del", s.taskName, queueBucket).Add(float64(n))
+	case ddl:
+		m.WithLabelValues("ddl", s.taskName, queueBucket).Add(float64(n))
+	case xid:
+		// ignore xid jobs
+	case flush:
+		m.WithLabelValues("flush", s.taskName, queueBucket).Add(float64(n))
+	case skip:
+		// ignore skip jobs
+	default:
+		log.Warnf("unknown optype %v", tp)
+	}
+
+	s.count.Add(n)
+}
+
+func (s *SyncPipe) checkWait(job *job) bool {
+	if job.tp == ddl {
+		return true
+	}
+
+	if s.checkpoint.CheckGlobalPoint() {
+		return true
+	}
+
+	return false
+}
+
+func (s *SyncPipe) addJob(job *job) error {
+	var (
+		queueBucket int
+		execDDLReq  *pb.ExecDDLRequest
+	)
+	switch job.tp {
+	case xid:
+		s.saveGlobalPoint(job.pos)
+		return nil
+	case flush:
+		addedJobsTotal.WithLabelValues("flush", s.taskName, adminQueueName).Inc()
+		// ugly code addJob and sync, refine it later
+		s.jobWg.Add(s.workerCount)
+		for i := 0; i < s.workerCount; i++ {
+			s.jobs[i] <- job
+		}
+		s.jobWg.Wait()
+		finishedJobsTotal.WithLabelValues("flush", s.taskName, adminQueueName).Inc()
+		return errors.Trace(s.flushCheckPoints())
+	case ddl:
+		s.jobWg.Wait()
+		addedJobsTotal.WithLabelValues("ddl", s.taskName, adminQueueName).Inc()
+		s.jobWg.Add(1)
+		queueBucket = s.workerCount
+		s.jobs[queueBucket] <- job
+		if job.ddlExecItem != nil {
+			execDDLReq = job.ddlExecItem.req
+		}
+
+	case insert, update, del:
+		s.jobWg.Add(1)
+		queueBucket = int(utils.GenHashKey(job.key)) % s.workerCount
+		s.addCount(false, s.queueBucketMapping[queueBucket], job.tp, 1)
+		s.jobs[queueBucket] <- job
+	}
+
+	if s.tracer.Enable() {
+		_, err := s.tracer.CollectSyncerJobEvent(job.traceID, job.traceGID, int32(job.tp), job.pos, job.currentPos, s.queueBucketMapping[queueBucket], job.sql, job.ddls, job.args, execDDLReq, pb.SyncerJobState_queued)
+		if err != nil {
+			log.Errorf("[syncer] trace error: %s", err)
+		}
+	}
+
+	wait := s.checkWait(job)
+	if wait {
+		s.jobWg.Wait()
+		s.c.reset()
+	}
+
+	switch job.tp {
+	case ddl:
+		// only save checkpoint for DDL and XID (see above)
+		s.saveGlobalPoint(job.pos)
+		if len(job.sourceSchema) > 0 {
+			log.Infof("save checkpoint %v, schema: %s, table: %s, pos: %v", s.checkpoint, job.sourceSchema, job.sourceTable, job.pos)
+			s.checkpoint.SaveTablePoint(job.sourceSchema, job.sourceTable, job.pos)
+		}
+		// reset sharding group after checkpoint saved
+		s.resetShardingGroup(job.targetSchema, job.targetTable)
+	case insert, update, del:
+		// save job's current pos for DML events
+		if len(job.sourceSchema) > 0 {
+			log.Infof("save checkpoint %v, schema: %s, table: %s, pos: %v", s.checkpoint, job.sourceSchema, job.sourceTable, job.currentPos)
+			s.checkpoint.SaveTablePoint(job.sourceSchema, job.sourceTable, job.currentPos)
+		}
+	}
+
+	if wait {
+		return errors.Trace(s.flushCheckPoints())
+	}
+
+	return nil
+}
+
+func (s *SyncPipe) saveGlobalPoint(globalPoint mysql.Position) {
```

Comment:
can we make this pipe thinner or split to more pipes? one pipe do one thing if possible?